provisioning/image_customization.go

```diff
@@ -0,0 +1,255 @@
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	metal3iov1alpha1 "github.com/openshift/cluster-baremetal-operator/api/v1alpha1"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+)
+
+const (
+	ironicBaseUrl                    = "IRONIC_BASE_URL"
+	ironicAgentImage                 = "IRONIC_AGENT_IMAGE"
+	imageCustomizationDeploymentName = "metal3-image-customization"
+	imageCustomizationVolume         = "metal3-image-customization-volume"
+	imageCustomizationPort           = 8084
+	containerRegistriesConfPath      = "/etc/containers/registries.conf"
+	containerRegistriesEnvVar        = "REGISTRIES_CONF_PATH"
+	deployISOEnvVar                  = "DEPLOY_ISO"
+	deployISOFile                    = "/shared/html/images/ironic-python-agent.iso"
+	deployInitrdEnvVar               = "DEPLOY_INITRD"
+	deployInitrdFile                 = "/shared/html/images/ironic-python-agent.initramfs"
+)
+
+var (
+	imageRegistriesVolumeMount = corev1.VolumeMount{
+		Name:      imageCustomizationVolume,
+		MountPath: containerRegistriesConfPath,
+	}
+)
+
+func imageRegistriesVolume() corev1.Volume {
+	// TODO: Should this be corev1.HostPathFile?
+	volType := corev1.HostPathFileOrCreate
+
+	return corev1.Volume{
+		Name: imageCustomizationVolume,
+		VolumeSource: corev1.VolumeSource{
+			HostPath: &corev1.HostPathVolumeSource{
+				Path: containerRegistriesConfPath,
+				Type: &volType,
+			},
+		},
+	}
+}
+
+func hostForIP(ipAddr string) string {
+	if strings.Contains(ipAddr, ":") {
+		return fmt.Sprintf("[%s]", ipAddr)
+	}
+	return ipAddr
+}
+
+func setIronicBaseUrl(name string, info *ProvisioningInfo) corev1.EnvVar {
+	config := info.ProvConfig.Spec
+	if config.ProvisioningNetwork != metal3iov1alpha1.ProvisioningNetworkDisabled && !config.VirtualMediaViaExternalNetwork {
+		return corev1.EnvVar{
+			Name:  name,
+			Value: "https://" + hostForIP(config.ProvisioningIP),
+		}
+	} else {
+		hostIP, err := getPodHostIP(info.Client.CoreV1(), info.Namespace)
+		if err == nil && hostIP != "" {
+			return corev1.EnvVar{
+				Name:  name,
+				Value: "https://" + hostForIP(hostIP),
+			}
+		}
+		return corev1.EnvVar{
+			Name: name,
+		}
+	}
+}
+
+func createImageCustomizationContainer(images *Images, info *ProvisioningInfo) corev1.Container {
+	container := corev1.Container{
+		Name:  "image-customization-controller",
+		Image: images.ImageCustomizationController,
+		Command: []string{"/image-customization-controller",
+			"-images-bind-addr", fmt.Sprintf(":%d", imageCustomizationPort),
+			"-images-publish-addr",
+			fmt.Sprintf("http://%s.%s.svc.cluster.local/",
+				imageCustomizationService, info.Namespace)},
+		SecurityContext: &corev1.SecurityContext{
+			Privileged: pointer.BoolPtr(true),
```

Comment:
This was worth a comment I thought, because we will eventually want to get rid of it when we give i-c-c its own volume.

Reply:
Done.